xtask/src/codegen/gen_lint_completions.rs

```diff
@@ -0,0 +1,115 @@
+    if !Path::new("./target/rust").exists() {
+        run!("git clone --depth=1 https://github.com/rust-lang/rust ./target/rust")?;
+    }
+
+    let ts_features = generate_descriptor("./target/rust/src/doc/unstable-book/src".into())?;
+    run!("curl http://rust-lang.github.io/rust-clippy/master/lints.json --output ./target/clippy_lints.json")?;
+
+    let ts_clippy = generate_descriptor_clippy("./target/clippy_lints.json")?;
+    let ts = quote! {
+        use crate::completion::complete_attribute::LintCompletion;
+        #ts_features
+        #ts_clippy
+    };
+    let contents = reformat(ts)?;
+
+    let destination =
+        project_root().join("crates/ide/src/completion/generated_lint_completions.rs");
+    update(destination.as_path(), &contents, mode)?;
+    run_rustfmt(mode)?;
+
+    Ok(())
+}
+
+fn generate_descriptor(src_dir: PathBuf) -> Result<proc_macro2::TokenStream> {
+    let definitions = ["language-features", "library-features"]
+        .iter()
+        .flat_map(|it| WalkDir::new(src_dir.join(it)))
+        .filter_map(|e| e.ok())
+        .filter(|entry| {
+            // Get all `.md ` files
+            entry.file_type().is_file() && entry.path().extension().unwrap_or_default() == "md"
+        })
+        .map(|entry| {
+            let path = entry.path();
+            let feature_ident = path.file_stem().unwrap().to_str().unwrap().replace("-", "_");
+            let doc = fs2::read_to_string(path).unwrap();
+
+            quote! { LintCompletion { label: #feature_ident, description: #doc } }
+        });
+
+    let ts = quote! {
+        pub(super) const FEATURES:  &[LintCompletion] = &[
+            #(#definitions),*
+        ];
+    };
+
+    Ok(ts)
+}
+
+#[derive(Default)]
+struct ClippyLint {
+    help: String,
+    id: String,
+}
+
+fn generate_descriptor_clippy(uri: &str) -> Result<proc_macro2::TokenStream> {
+    let mut file = File::open(uri)?;
+    let mut file_content = String::new();
+    file.read_to_string(&mut file_content)?;
+    let mut clippy_lints: Vec<ClippyLint> = vec![];
+
+    for line in file_content.lines().map(|line| line.trim()) {
+        if line.starts_with(r#""id":"#) {
+            let clippy_lint = ClippyLint {
+                id: line
+                    .strip_prefix(r#""id": ""#)
+                    .expect("should be prefixed by id")
+                    .strip_suffix(r#"","#)
+                    .expect("should be suffixed by comma")
+                    .into(),
+                help: String::new(),
+            };
+            clippy_lints.push(clippy_lint)
+        } else if line.starts_with(r#""What it does":"#) {
+            clippy_lints.last_mut().expect("clippy lint must already exist").help = line
+                .strip_prefix(r#""What it does": ""#)
+                .expect("should be prefixed by what it does")
+                .strip_suffix(r#"","#)
+                .expect("should be suffixed by comma")
+                .into();
```

Comment:
this is not very readable, readability can be improved by extracing right hand side into a variable.

Reply:
bump